### PR TITLE
Update jetpack masterbar design

### PR DESCRIPTION
**Pull request by CodeyGuyDylan**
## Proposed Changes

* Update masterbar design to match the new comps
* Fix issue where the "backdrop" was able to be scrolled past resulting in this
image
* Fix issue where logo and gravatar was not showing when products submenu was open (it is on jetpack.com)
* Makes header sticky like on jetpack.com instead of these things
image
* Make sure the user menu (when logged in) still works as expected
* Make sure the mobile menu works as expected.
* Try opening in incognito initially and making sure both desktop and mobile still work
* Make sure clicking on the blur background closes the megamenu
* Log out and make sure all menus still work as expected

Figma: 22h2u6LRIyamGpA0vEo6hW-fi-3708%3A657
P2: p1HpG7-nDJ-p2

## Testing Instructions

1. Click on the Calypso Green live link or checkout this branch and start your local environment via `yarn start-jetpack-cloud`
2. Go to `/pricing`
3. Make sure the header looks and functions exactly the same on mobile and desktop as it does in this diff: D118381-code as well as looks the same as in the Figma comps
4. Make sure none of the fixes listed in the Proposed changes broke anything else on the header or caused any unexpected issues

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by CodeyGuyDylan** [comment]
FYI this includes changes from https://github.com/Automattic/wp-calypso/pull/80241, will rebased once that is approved and merged

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js
Comment (robertsreberski) [inline client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js]: I need help understanding this comment with the combination that you removed some code from below. Was this code troublesome, or is it solved somewhere else?

File: client/data/jetpack/use-jetpack-masterbar-data-query.ts
Comment (robertsreberski) [inline client/data/jetpack/use-jetpack-masterbar-data-query.ts]: Question: are we making sure that this data is loaded before we show the page? Or do we present some kind of loader if it takes longer? And what in case of error coming from this endpoint?

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (robertsreberski) [inline client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:58]: I think something is still wrong with order - I suppose "Products" should be first in the menu, analogically to jetpack.com menu:
CleanShot 2023-08-14 at 16 53 29@2x

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (robertsreberski) [inline client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:117]: In general, the component code and styles look good when tested - similar to jetpack.com diff.

However, I'm wondering if we could improve the readability of this code. React gives us a nice possibility to modularize view components, and here we end up with a large nested structure that has many indents.

We could at least make a small improvement by setting a new pattern in the new code that is added here - I don't mean the big refactor of the whole file at once. Making such a big change to the menu is a good opportunity to do that.

**Review by robertsreberski — COMMENTED** [review]
The code tests well, but I suggest it could be made more readable. That's not a blocker, though and I understand if you decide to go with it as-is.

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js
Comment (CodeyGuyDylan) [reply client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js]: There was a function that handled opening an closing the menu here, it was causing issues because this script was getting loaded multiple times due to the new data query and the event listeners were being added multiple times. I moved this to an onClick function in the React component.

File: client/data/jetpack/use-jetpack-masterbar-data-query.ts
Comment (CodeyGuyDylan) [reply client/data/jetpack/use-jetpack-masterbar-data-query.ts]: In the case of an error or no data being loaded, the menu will just be the jetpack logo and the user gravatar on the right hand side. While the data is loaded, that is also what is being shown

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (CodeyGuyDylan) [reply client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:58]: That is strange, the menu_order of Products is correct, and I have only ever seen it be the first item in the menu since I've sorted it that way. I cannot reproduce this but it is concerning that you see it that way 🤔

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (CodeyGuyDylan) [reply client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:117]: I have a separate task to move some of this to components and clean up the code 😄

**Comment by CodeyGuyDylan** [comment]
I also discovered a bug where the products menu does not open the first time the menu data is queried by the user, looking into that

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (CodeyGuyDylan) [reply client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:58]: I also used the `/fr/` localizer to get the GDPR link to show up, but it is still in the correct order for me 🤔

**Comment by CodeyGuyDylan** [comment]
> I also discovered a bug where the products menu does not open the first time the menu data is queried by the user, looking into that

The issue is the submenu.js file add's a `js` class to all menu buttons, but since the data takes a second or two to load the first time, those elements do not exist when the JavaScript tries to do that. I am removing this functionality and adding the `js` class and the initial `hidden` value directly to the markup instead. As soon as this PR gets merged, I am going to start working on cleaning up the React in general

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js
Comment (jboland88) [inline client/jetpack-cloud/sections/pricing/jpcom-masterbar/use-submenu-btn.js:39]: I think you might still need this bit since the rest of the code assumes the menu is defined?

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (jboland88) [inline client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:126]: What would happen if you called these functions inside your new `useEffect` statement? 
Not sure if React would throw an error since these functions also employ `useEffect` 🤔 

That could allow for some of the code to stay separate/ more readable for now?

**Review by jboland88 — APPROVED** [review]
Overall, this tested well.

Code looks fine for the current iteration - I had one idea about maybe keeping the event handlers in the separate files for now. If we do leave the event handlers all in the jpcom-masterbar component like they are here, let's add some comments to note why it is like this for now.

I did find a few small issues, let's get the first one fixed before merging: 
- I can't fully close the mobile menu if the products tray is open. Clicking the close menu button in this case just closes the product tray.
- Product bundles header animates in from the right by itself in the mobile view.
- If mobile menu is still "open" and you scale back up to desktop, the screen is still locked and does not scroll.

Approving with the caveat of fixing the mobile menu button issue

File: client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx
Comment (CodeyGuyDylan) [reply client/jetpack-cloud/sections/pricing/jpcom-masterbar/index.tsx:126]: Yeah I tried that initially, but React doesn’t allow useEffect to be called within any sort of callback. This is the specific error I get
`React Hook "useEffect" cannot be called inside a callback. React Hooks must be called in a React function component or a custom React Hook function.`

**Pull request merged**
